<a href="https://colab.research.google.com/github/AlienX77-cmd/Python-for-Business-and-Finance/blob/main/Part_3_Exercise_01.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Coding Exercises (Part 3)

Now, you will have the opportunity to practice what you have learned. <br>
__Follow the instructions__ and insert your code! 

The correct answer is provided below your coding cell. There you can check whether your code is correct.

If you need some help or if you want to check your code, you can also have a look at the __solutions__.

### Have Fun!

--------------------------------------------------------------------------------------------------------------

## Exercise 1: N-Dimensional Numpy Arrays

In [1]:
# run the cell!
import numpy as np
np.set_printoptions(precision=2, suppress= True)

Let´s assume we have three investment projects with the following cashflow plans:

In [2]:
# run the cell!
project_1 = [-500, 100, 150, 200, 200, 175, 100, 50]
project_2 = [-750, 100, 125, 150, 175, 200, 250, 0]
project_3 = [-2000, 500, 1000, 2500, 0, 0, 0, 0]

1. Create and save a __2-dimensional Numpy Array__ (in the variable __projects__).

In [7]:
projects = np.array([project_1, project_2, project_3])

In [8]:
projects

array([[ -500,   100,   150,   200,   200,   175,   100,    50],
       [ -750,   100,   125,   150,   175,   200,   250,     0],
       [-2000,   500,  1000,  2500,     0,     0,     0,     0]])

2. Check the __shape__ of projects. __How many rows/columns__ do we have?

In [10]:
projects.shape

(3, 8)

(result: 3 rows, 8 columns)

3. __Slice__ projects for __the second project__.

In [14]:
projects[1, :]

array([-750,  100,  125,  150,  175,  200,  250,    0])

4. Return the __initial Investment__ of __Project 3__.

In [15]:
projects[2,0]

-2000

(result: -2000)

5. Return the __final cashflow__ in year 7 for __all projects__.

In [16]:
projects[:, 7]

array([50,  0,  0])

6. Return the first positive cashflows (__Year 1__) for __Projects 1 and 2__.

In [20]:
projects[0:2, 1]

array([100, 100])

(result: array([100, 100]))

7. __Change__ the initial investment for __project 3__ to __-1900__.

In [21]:
projects[2, 0] = -1900

In [22]:
projects

array([[ -500,   100,   150,   200,   200,   175,   100,    50],
       [ -750,   100,   125,   150,   175,   200,   250,     0],
       [-1900,   500,  1000,  2500,     0,     0,     0,     0]])

8. Change the __last cashflows__ in year 7 for all projects to __25__.

In [23]:
projects[:, 7] = 25

In [24]:
projects

array([[ -500,   100,   150,   200,   200,   175,   100,    25],
       [ -750,   100,   125,   150,   175,   200,   250,    25],
       [-1900,   500,  1000,  2500,     0,     0,     0,    25]])

9. Calculate the total cashflows on each timestamp over all three projects (__column-wise sum__).

In [25]:
projects.sum(axis = 0)

array([-3150,   700,  1275,  2850,   375,   375,   350,    75])

(results: array([-3150,   700,  1275,  2850,   375,   375,   350,    75]))

10. Calculate __cumulative cashflows__ for all three projects.

In [27]:
projects.cumsum(axis = 1)

array([[ -500,  -400,  -250,   -50,   150,   325,   425,   450],
       [ -750,  -650,  -525,  -375,  -200,     0,   250,   275],
       [-1900, -1400,  -400,  2100,  2100,  2100,  2100,  2125]])

11. Convert __all projects__ to EUR currency (__multiply with 0.9__).

In [29]:
projects * 0.9

array([[ -450. ,    90. ,   135. ,   180. ,   180. ,   157.5,    90. ,
           22.5],
       [ -675. ,    90. ,   112.5,   135. ,   157.5,   180. ,   225. ,
           22.5],
       [-1710. ,   450. ,   900. ,  2250. ,     0. ,     0. ,     0. ,
           22.5]])

12. Convert __Project 1__ to EUR currency (__multiply with 0.9__), __Project 2__ to GBP currency (__multiply with 0.77__) and __Project 3__ to Yen Currency (__multiply with 109__).

In [31]:
mult = np.array([0.9, 0.77, 109]).reshape(3, 1)
mult

array([[  0.9 ],
       [  0.77],
       [109.  ]])

In [32]:
projects * mult

array([[   -450.  ,      90.  ,     135.  ,     180.  ,     180.  ,
            157.5 ,      90.  ,      22.5 ],
       [   -577.5 ,      77.  ,      96.25,     115.5 ,     134.75,
            154.  ,     192.5 ,      19.25],
       [-207100.  ,   54500.  ,  109000.  ,  272500.  ,       0.  ,
              0.  ,       0.  ,    2725.  ]])

13. __Transpose__ projects!

In [35]:
projects.T

array([[ -500,  -750, -1900],
       [  100,   100,   500],
       [  150,   125,  1000],
       [  200,   150,  2500],
       [  200,   175,     0],
       [  175,   200,     0],
       [  100,   250,     0],
       [   25,    25,    25]])

14. Let´s assume that all three projects are postponed by one year. "Add" a __column with zero__ initial cashflows.

In [40]:
zeros = np.zeros(3).reshape(3,1)

In [41]:
np.hstack([zeros, projects])

array([[    0.,  -500.,   100.,   150.,   200.,   200.,   175.,   100.,
           25.],
       [    0.,  -750.,   100.,   125.,   150.,   175.,   200.,   250.,
           25.],
       [    0., -1900.,   500.,  1000.,  2500.,     0.,     0.,     0.,
           25.]])

15. __Simulate__ playing Roulette 10 times (equally likely outcomes from __0 until 36__). Use a __random seed of 123__. 

In [42]:
np.random.seed(123)
np.random.randint(0, 37, 10)

array([ 2, 28, 34, 17, 19, 22, 33, 32,  9, 32])

16. Same as 15. but 100,000 simulations (__100,000 times 10 games__). Save the array in the variable roulette. Use a __random seed of 123__.

In [43]:
sims = 100000
games = 10

In [44]:
np.random.seed(123)
roulette = np.random.randint(0, 37, sims*games).reshape(sims, games)

In [45]:
roulette

array([[ 2, 28, 34, ..., 32,  9, 32],
       [32, 25, 19, ...,  4,  3,  2],
       [20,  2, 20, ..., 21, 30, 27],
       ...,
       [11,  3, 34, ..., 32, 15, 22],
       [18,  8, 16, ..., 35, 13, 32],
       [22,  2, 34, ..., 20, 24,  7]])

17. In how many simulations do we get the __number 13__ __four times__?

In [46]:
thirteen = (roulette == 13).sum(axis = 1)
thirteen[thirteen == 4].size

8

(result: 8)

# Well Done!

---------------------------------------------------------------------------------------------------------------------

# Solutions (Stop here if you want to code on your own!)

# ++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++

In [ ]:
# run the cell!
import numpy as np
np.set_printoptions(precision=2, suppress= True)

Let´s assume we have three investment projects with the following cashflow plans:

In [ ]:
# run the cell!
project_1 = [-500, 100, 150, 200, 200, 175, 100, 50]
project_2 = [-750, 100, 125, 150, 175, 200, 250, 0]
project_3 = [-2000, 500, 1000, 2500, 0, 0, 0, 0]

1. Create and save a __2-dimensional Numpy Array__ (in the variable __projects__).

In [ ]:
projects = np.array([project_1, project_2, project_3])

In [ ]:
projects

2. Check the __shape__ of projects. __How many rows/columns__ do we have?

In [ ]:
projects.shape

(result: 3 rows, 8 columns)

3. __Slice__ projects for __the second project__.

In [ ]:
projects[1]

4. Return the __initial Investment__ of __Project 3__.

In [ ]:
projects[-1, 0]

(result: -2000)

5. Return the __final cashflow__ in year 7 for __all projects__.

In [ ]:
projects[:, -1]

6. Return the first positive cashflows (__Year 1__) for __Projects 1 and 2__.

In [ ]:
projects[:2, 1]

(result: array([100, 100]))

7. __Change__ the initial investment for __project 3__ to __-1900__.

In [ ]:
projects[-1, 0] = -1900

In [ ]:
projects

8. Change the __last cashflows__ in year 7 for all projects to __25__.

In [ ]:
projects[:, -1] = 25

In [ ]:
projects

9. Calculate the total cashflows on each timestamp over all three projects (__column-wise sum__).

In [ ]:
projects.sum(axis = 0)

(results: array([-3150,   700,  1275,  2850,   375,   375,   350,    75]))

10. Calculate __cumulative cashflows__ for all three projects.

In [ ]:
projects.cumsum(axis = 1)

11. Convert __all projects__ to EUR currency (__multiply with 0.9__).

In [ ]:
projects * 0.9

12. Convert __Project 1__ to EUR currency (__multiply with 0.9__), __Project 2__ to GBP currency (__multiply with 0.77__) and __Project 3__ to Yen Currency (__multiply with 109__).

In [ ]:
vector = np.array([0.9, 0.77, 109]).reshape(3,1)
vector

In [ ]:
projects * vector

13. __Transpose__ projects!

In [ ]:
projects.T

In [ ]:
projects.transpose()

14. Let´s assume that all three projects are postponed by one year. "Add" a __column with zero__ initial cashflows.

In [ ]:
zeros = np.zeros(3).reshape(3,1)

In [ ]:
np.hstack([zeros, projects])

15. __Simulate__ playing Roulette 10 times (equally likely outcomes from __0 until 36__). Use a __random seed of 123__. 

In [ ]:
np.random.seed(123)
np.random.randint(0, 37, 10)

16. Same as 15. but 100,000 simulations (__100,000 times 10 games__). Save the array in the variable roulette.

In [ ]:
games = 10
sims = 100000

In [ ]:
np.random.seed(123)
roulette = np.random.randint(0, 37, sims * games).reshape(sims, games)

17. In how many simulations do we get the __number 13__ __four times__?

In [ ]:
thirteen = (roulette == 13).sum(axis = 1)

In [ ]:
thirteen[thirteen == 4].size

(result: 8)